File: internal/paths/paths_test.go
package paths_test

import (
	"errors"
	"testing"

	"github.com/huaiche94/preflight/internal/paths"
)

// --- Windows/macOS/Linux path-table tests (agents/foundation.md "Required
// tests" bullet) -----------------------------------------------------------

func TestResolve_Linux_XDGDefaults(t *testing.T) {
	env := newFakeEnv("/home/alice")

	dirs, err := paths.Resolve("linux", env)
	if err != nil {
		t.Fatalf("Resolve: %v", err)
	}

	want := paths.Dirs{
		Config:  "/home/alice/.config/preflight",
		Data:    "/home/alice/.local/share/preflight",
		Cache:   "/home/alice/.cache/preflight",
		Runtime: "/home/alice/.cache/preflight/run",
	}
	assertDirsEqual(t, dirs, want)
}

func TestResolve_Linux_XDGOverrides(t *testing.T) {
	env := newFakeEnv("/home/alice").
		with("XDG_CONFIG_HOME", "/custom/config").
		with("XDG_DATA_HOME", "/custom/data").
		with("XDG_CACHE_HOME", "/custom/cache").
		with("XDG_RUNTIME_DIR", "/run/user/1000")

	dirs, err := paths.Resolve("linux", env)
	if err != nil {
		t.Fatalf("Resolve: %v", err)
	}

	want := paths.Dirs{
		Config:  "/custom/config/preflight",
		Data:    "/custom/data/preflight",
		Cache:   "/custom/cache/preflight",
		Runtime: "/run/user/1000/preflight",
	}
	assertDirsEqual(t, dirs, want)
}

func TestResolve_FreeBSD_TreatedAsXDG(t *testing.T) {
	// Any non-windows/non-darwin GOOS should fall through to the XDG
	// resolver, since Preflight's portability goal is POSIX-general, not
	// Linux-specific.
	env := newFakeEnv("/home/bob")

	dirs, err := paths.Resolve("freebsd", env)
	if err != nil {
		t.Fatalf("Resolve: %v", err)
	}
	if dirs.Config != "/home/bob/.config/preflight" {
		t.Errorf("Config = %q, want XDG default", dirs.Config)
	}
}

func TestResolve_Darwin_Defaults(t *testing.T) {
	env := newFakeEnv("/Users/alice")

	dirs, err := paths.Resolve("darwin", env)
	if err != nil {
		t.Fatalf("Resolve: %v", err)
	}

	want := paths.Dirs{
		Config:  "/Users/alice/Library/Application Support/preflight",
		Data:    "/Users/alice/Library/Application Support/preflight",
		Cache:   "/Users/alice/Library/Caches/preflight",
		Runtime: "/Users/alice/Library/Caches/preflight/run",
	}
	assertDirsEqual(t, dirs, want)
}

func TestResolve_Windows_Defaults(t *testing.T) {
	env := newFakeEnv(`C:\Users\alice`).
		with("APPDATA", `C:\Users\alice\AppData\Roaming`).
		with("LOCALAPPDATA", `C:\Users\alice\AppData\Local`)

	dirs, err := paths.Resolve("windows", env)
	if err != nil {
		t.Fatalf("Resolve: %v", err)
	}

	want := paths.Dirs{
		Config:  `C:\Users\alice\AppData\Roaming\preflight\Config`,
		Data:    `C:\Users\alice\AppData\Local\preflight\Data`,
		Cache:   `C:\Users\alice\AppData\Local\preflight\Cache`,
		Runtime: `C:\Users\alice\AppData\Local\preflight\Run`,
	}
	assertDirsEqual(t, dirs, want)
}

func TestResolve_Windows_FallsBackToHomeWhenEnvUnset(t *testing.T) {
	env := newFakeEnv(`C:\Users\alice`)

	dirs, err := paths.Resolve("windows", env)
	if err != nil {
		t.Fatalf("Resolve: %v", err)
	}

	const wantConfig = `C:\Users\alice\AppData\Roaming\preflight\Config`
	if dirs.Config != wantConfig {
		t.Errorf("Config = %q, want %q", dirs.Config, wantConfig)
	}
}

// --- Injectable environment/home behavior ----------------------------------

func TestResolve_MissingHomeDir_ReturnsError(t *testing.T) {
	env := &fakeEnv{vars: map[string]string{}, homeErr: errors.New("no home")}

	if _, err := paths.Resolve("linux", env); err == nil {
		t.Fatal("expected error when home directory cannot be determined")
	} else if !errors.Is(err, paths.ErrNoHomeDir) {
		t.Errorf("error = %v, want wrapped ErrNoHomeDir", err)
	}
}

func TestResolve_EmptyHomeDir_ReturnsError(t *testing.T) {
	env := newFakeEnv("")

	if _, err := paths.Resolve("linux", env); !errors.Is(err, paths.ErrNoHomeDir) {
		t.Errorf("error = %v, want ErrNoHomeDir", err)
	}
}

func TestResolve_Windows_DoesNotRequireHomeWhenEnvSet(t *testing.T) {
	// APPDATA/LOCALAPPDATA fully set means UserHomeDir should never be
	// consulted; simulate an environment where it would error if called.
	env := &fakeEnv{
		vars: map[string]string{
			"APPDATA":      `C:\Users\alice\AppData\Roaming`,
			"LOCALAPPDATA": `C:\Users\alice\AppData\Local`,
		},
		homeErr: errors.New("home should not be consulted"),
	}

	if _, err := paths.Resolve("windows", env); err != nil {
		t.Fatalf("Resolve: %v", err)
	}
}

// --- ResolveHost / real OSEnv smoke test -----------------------------------

func TestResolveHost_UsesRealEnv(t *testing.T) {
	dirs, err := paths.ResolveHost(paths.NewOSEnv())
	if err != nil {
		t.Fatalf("ResolveHost: %v", err)
	}
	if dirs.Config == "" || dirs.Data == "" || dirs.Cache == "" || dirs.Runtime == "" {
		t.Errorf("ResolveHost returned an empty directory: %+v", dirs)
	}
}

func assertDirsEqual(t *testing.T, got, want paths.Dirs) {
	t.Helper()
	if got != want {
		t.Errorf("Dirs mismatch:\n got  = %+v\n want = %+v", got, want)
	}
}

File: vscode/src/test/paths.test.ts
/**
 * paths.test.ts — per-OS directory resolution with injected env, proving
 * the TypeScript mirror agrees with internal/paths/paths.go on every OS
 * branch (the same all-branches-from-one-host strategy paths.go's own
 * tests use via their injectable Env).
 */
import assert from 'node:assert/strict';
import { test } from 'node:test';

import { Env, NoHomeDirError, resolveDirs, winJoin } from '../paths';

function fakeEnv(vars: Record<string, string>, home = '/home/alice'): Env {
  return {
    getenv: (key) => vars[key] ?? '',
    homedir: () => home,
  };
}

test('darwin mirrors paths.go resolveDarwin', () => {
  const dirs = resolveDirs('darwin', fakeEnv({}, '/Users/alice'));
  assert.equal(dirs.config, '/Users/alice/Library/Application Support/auspex');
  assert.equal(dirs.data, '/Users/alice/Library/Application Support/auspex');
  assert.equal(dirs.cache, '/Users/alice/Library/Caches/auspex');
  assert.equal(dirs.runtime, '/Users/alice/Library/Caches/auspex/run');
});

test('linux XDG defaults mirror paths.go resolveXDG', () => {
  const dirs = resolveDirs('linux', fakeEnv({}));
  assert.equal(dirs.config, '/home/alice/.config/auspex');
  assert.equal(dirs.data, '/home/alice/.local/share/auspex');
  assert.equal(dirs.cache, '/home/alice/.cache/auspex');
  // No XDG_RUNTIME_DIR: falls back to <cache base>/auspex/run.
  assert.equal(dirs.runtime, '/home/alice/.cache/auspex/run');
});

test('linux honors XDG overrides including XDG_RUNTIME_DIR', () => {
  const dirs = resolveDirs(
    'linux',
    fakeEnv({
      XDG_CONFIG_HOME: '/xdg/config',
      XDG_DATA_HOME: '/xdg/data',
      XDG_CACHE_HOME: '/xdg/cache',
      XDG_RUNTIME_DIR: '/run/user/1000',
    })
  );
  assert.equal(dirs.config, '/xdg/config/auspex');
  assert.equal(dirs.data, '/xdg/data/auspex');
  assert.equal(dirs.cache, '/xdg/cache/auspex');
  assert.equal(dirs.runtime, '/run/user/1000/auspex');
});

test('empty env values are treated as unset (paths.go Getenv contract)', () => {
  const dirs = resolveDirs('linux', fakeEnv({ XDG_DATA_HOME: '', XDG_RUNTIME_DIR: '' }));
  assert.equal(dirs.data, '/home/alice/.local/share/auspex');
  assert.equal(dirs.runtime, '/home/alice/.cache/auspex/run');
});

test('non-darwin non-windows platforms take the XDG branch', () => {
  // paths.go treats every other GOOS as XDG-conformant POSIX.
  const dirs = resolveDirs('freebsd', fakeEnv({}));
  assert.equal(dirs.data, '/home/alice/.local/share/auspex');
});

test('win32 mirrors paths.go resolveWindows with env set', () => {
  const dirs = resolveDirs(
    'win32',
    fakeEnv({
      APPDATA: 'C:\\Users\\alice\\AppData\\Roaming',
      LOCALAPPDATA: 'C:\\Users\\alice\\AppData\\Local',
    })
  );
  assert.equal(dirs.config, 'C:\\Users\\alice\\AppData\\Roaming\\auspex\\Config');
  assert.equal(dirs.data, 'C:\\Users\\alice\\AppData\\Local\\auspex\\Data');
  assert.equal(dirs.cache, 'C:\\Users\\alice\\AppData\\Local\\auspex\\Cache');
  assert.equal(dirs.runtime, 'C:\\Users\\alice\\AppData\\Local\\auspex\\Run');
});

test('win32 falls back to home AppData when env unset', () => {
  const dirs = resolveDirs('win32', fakeEnv({}, 'C:\\Users\\alice'));
  assert.equal(dirs.config, 'C:\\Users\\alice\\AppData\\Roaming\\auspex\\Config');
  assert.equal(dirs.runtime, 'C:\\Users\\alice\\AppData\\Local\\auspex\\Run');
});

test('winJoin trims stray backslashes like paths.go winJoin', () => {
  assert.equal(winJoin('C:\\base\\', 'auspex', 'Data'), 'C:\\base\\auspex\\Data');
  assert.equal(winJoin('', 'auspex'), 'auspex');
});

test('missing home directory throws NoHomeDirError', () => {
  assert.throws(() => resolveDirs('linux', fakeEnv({}, '')), NoHomeDirError);
  assert.throws(() => resolveDirs('darwin', fakeEnv({}, '')), NoHomeDirError);
  // windows only needs home when APPDATA/LOCALAPPDATA are unset:
  assert.throws(() => resolveDirs('win32', fakeEnv({}, '')), NoHomeDirError);
  const dirs = resolveDirs('win32', fakeEnv({ APPDATA: 'C:\\a', LOCALAPPDATA: 'C:\\b' }, ''));
  assert.equal(dirs.data, 'C:\\b\\auspex\\Data');
});

test('trailing slashes in XDG values collapse like Go path.Join', () => {
  const dirs = resolveDirs('linux', fakeEnv({ XDG_DATA_HOME: '/xdg/data/' }));
  assert.equal(dirs.data, '/xdg/data/auspex');
});
